File: src/math/__validation__/drdidValidation.js
// ─── ECON STUDIO · src/math/__validation__/drdidValidation.js ─────────────────
// Validation harness for compute2x2 (doubly-robust 2×2 DiD).
// Franco pastes R fixtures from drdidValidation.R into R_FIXTURES below.
// Then: window.__validation.drdid shows pass/total in browser console.

import { compute2x2 } from "../did/drdid.js";

const SAMPLE = {
  deltaY: [1.9, 2.1, 1.7, 2.3, 2.0, 1.8,  0.4, 0.6, 0.5, 0.3, 0.7, 0.5],
  D:      [1,   1,   1,   1,   1,   1,     0,   0,   0,   0,   0,   0  ],
  X:      [
    [1, 0.2], [1, 0.5], [1,-0.1], [1, 0.8], [1, 0.3], [1, 0.0],
    [1, 1.1], [1, 0.9], [1, 1.4], [1, 1.0], [1, 1.2], [1, 0.7],
  ],
};
const n = SAMPLE.deltaY.length;
const weights = Array(n).fill(1);

// R fixtures — Franco pastes output from drdidValidation.R here:
// e.g.: R_FIXTURES["reg"] = { att: 1.700000, se: 0.058926 };
const R_FIXTURES = {
  reg: { att: NaN, se: NaN },
  ipw: { att: NaN, se: NaN },
  dr:  { att: NaN, se: NaN },
};

export function runDrdidValidation() {
  const rows = [];
  for (const m of ["reg", "ipw", "dr"]) {
    const result = compute2x2({ ...SAMPLE, estMethod: m, weights });
    const { att, inf } = result;

    // se = sqrt(mean(inf²)) / n   — matches DRDID R package convention
    let s2 = 0;
    for (let i = 0; i < n; i++) s2 += inf[i] * inf[i];
    const se = Math.sqrt(s2 / n) / n;

    const rf = R_FIXTURES[m];
    const okAtt = !isNaN(rf.att) ? Math.abs(att - rf.att) < 1e-4 : null;
    const okSE  = !isNaN(rf.se)  ? Math.abs(se  - rf.se)  < 1e-4 : null;

    rows.push({
      method: m,
      att: +att.toFixed(6),
      attR: rf.att,
      diffAtt: !isNaN(rf.att) ? +(Math.abs(att - rf.att)).toFixed(6) : null,
      se: +se.toFixed(6),
      seR: rf.se,
      diffSE: !isNaN(rf.se) ? +(Math.abs(se - rf.se)).toFixed(6) : null,
      okAtt,
      okSE,
      warning: result.warning ?? null,
    });
  }
  console.table(rows);

  if (typeof window !== "undefined") {
    window.__validation = window.__validation ?? {};
    window.__validation.drdid = {
      rows,
      pass: rows.filter(r => r.okAtt === true && r.okSE === true).length,
      total: rows.length,
      noFixture: rows.filter(r => r.okAtt === null || r.okSE === null).length,
    };
  }
  return rows;
}

// Allow running directly: node src/math/__validation__/drdidValidation.js
if (typeof process !== "undefined" && process.argv?.[1]?.includes("drdidValidation")) {
  runDrdidValidation();
}

File: src/math/__validation__/drdidValidation.R
# ─── ECON STUDIO · src/math/__validation__/drdidValidation.R ──────────────────
# Generates R fixtures for the JS drdidValidation.js harness.
#
# Run:  Rscript src/math/__validation__/drdidValidation.R
# Requires: install.packages("DRDID")
#
# Paste the printed R_FIXTURES lines into drdidValidation.js.

library(DRDID)

deltaY <- c(1.9, 2.1, 1.7, 2.3, 2.0, 1.8,  0.4, 0.6, 0.5, 0.3, 0.7, 0.5)
D      <- c(rep(1L, 6), rep(0L, 6))
covX   <- c(0.2, 0.5, -0.1, 0.8, 0.3, 0.0,  1.1, 0.9, 1.4, 1.0, 1.2, 0.7)

# DRDID panel functions expect y1 (post) and y0 (pre).
# Here deltaY = y1 - y0, so we pass y1 = deltaY, y0 = 0.
y1 <- deltaY
y0 <- rep(0, length(deltaY))
covariates <- cbind(1, covX)   # n×2 matrix with intercept

n <- length(deltaY)

cat("// Paste these lines into R_FIXTURES in drdidValidation.js:\n")
for (m in c("reg", "ipw", "dr")) {
  fn <- switch(m,
    reg = reg_did_panel,
    ipw = std_ipw_did_panel,
    dr  = drdid_panel
  )
  res <- fn(y1 = y1, y0 = y0, D = D, covariates = covariates, inffunc = TRUE)
  se  <- sqrt(mean(res$att.inf.func^2)) / n
  cat(sprintf('  R_FIXTURES["%s"] = { att: %.6f, se: %.6f };\n', m, res$ATT, se))
}

cat("\n// Full output for inspection:\n")
for (m in c("reg", "ipw", "dr")) {
  fn <- switch(m,
    reg = reg_did_panel,
    ipw = std_ipw_did_panel,
    dr  = drdid_panel
  )
  res <- fn(y1 = y1, y0 = y0, D = D, covariates = covariates, inffunc = TRUE)
  se  <- sqrt(mean(res$att.inf.func^2)) / n
  cat(sprintf("\n[%s] ATT = %.6f  SE = %.6f\n", m, res$ATT, se))
  cat("  inf.func (first 6):", paste(round(res$att.inf.func[1:6], 6), collapse = ", "), "\n")
}
